### ISSUE TITLE
zfs changes back to /dev/sdx device names upon reboot

### ISSUE THREAD

**Issue by rpdrewes** [zfsonlinux/zfs#2944]
I use /dev/disk/by-id/ device names, but every time I reboot the devices are reimported as /dev/sdx names.  I export the tank, then force an import with the desired device names like this:

zpool import -d /dev/disk/by-id tank

and then zpool status shows the devices imported with the /dev/disk/by-id names.  I can then export the tank, and reimport with no -d argument and the devices come in with the /dev/disk/by-id names as they should.  However as soon as I reboot, the tank comes in with the /dev/sdx names that I don't want.  I suspect this has something to do with systemd and associated scripts, which is why I am reporting this issue here rather than to zfsonlinux.  Prior to upgrading to 14.04 I did not have this problem (though before the upgrade I did have problems getting the zfs tank to automount on boot, despite putting an extra delay in a systemd script; I had to finally edit /etc/defaul/zfs to force mount at boot, but those options were removed from /etc/default/zfs at the last upgrade and no longer seem to be necessary).

I have tried exporting, removing zpool.cache, reimporting with desired device names, and then rebooting, and the tank comes in with the undesired /dev/sdx names.

Thanks for the excellent work on zfs.  It has worked great for me for years.

Version info below.

Thanks,
Rich

root@main:~# dpkg -l | grep zfs
ii  dkms                                                  2.2.0.3-1.1ubuntu5+zfs9~trusty                      all          Dynamic Kernel Module Support Framework
ii  libzfs2                                               0.6.3-3~trusty                                      amd64        Native ZFS filesystem library for Linux
ii  mountall                                              2.53-zfs1                                           amd64        filesystem mounting tool
ii  ubuntu-zfs                                            8~trusty                                            amd64        Native ZFS filesystem metapackage for Ubuntu.
ii  zfs-dkms                                              0.6.3-3~trusty                                      amd64        Native ZFS filesystem kernel modules for Linux
ii  zfs-doc                                               0.6.3-3~trusty                                      amd64        Native ZFS filesystem documentation and examples.
ii  zfsutils                                              0.6.3-3~trusty                                      amd64        Native ZFS management utilities for Linux
root@main:~# uname -a
Linux main 3.13.0-40-generic #69-Ubuntu SMP Thu Nov 13 17:53:56 UTC 2014 x86_64 x86_64 x86_64 GNU/Linux

**Comment by Bronek** [comment]
I still have this problem, now using 0.6.3-1.2 plus small selection of patches merged from master zfs. I think this might be related to errors I often receive when running zpool import -d /dev/disk/by-id (it's a dual boot machine, both running exactly same kernel, zfs and spl binaries).

**Comment by draffelt** [comment]
Hi,
I am having the same issue on our 14.04 server. We have 15 drives in a pool, and every time I reboot different numbers of drives are labelled as either the /dev/sdX name or /dev/disk/by-id/. The actual drives that are labelled correctly by-id differ every time I reboot. Most drives are connected via a LSI SAS 9211-8i cards that are in JBOD mode (flashed with IT firmware), using the P20 version and associated linux drivers. The remaining drives are connected directly to an Intel SP2600COE motherboard in RSTe (pass through) mode. 

Also, about 1/10 boots the system hangs shortly after the ZFS module is loaded. If I export the pool and disconnect the drives it boots successfully 100% of the time. 

I'm guessing its an issue related to the order everything is loaded. Any help or suggestions would be much appreciated.  Not sure if this is the best place to ask for help, however it seems to be related to the issues others have been having.  I'm happy to email the boot logs if it helps.
Cheers,
Dave

**Comment by draffelt** [comment]
I've followed the instructions to add a [devicewait script](https://github.com/zfsonlinux/pkg-zfs/wiki/Ubuntu-ZFS-mountall-FAQ-and-troubleshooting) to mountall.conf. So far it appears to have fixed the device renaming, however it still hangs on mount about 3/10 times (even though the devicewait script successfully finds all drives in the pool).

The only other clue is that when I hit ctrl-alt-delete after it hangs, it outputs a message to the screen "An error occurred mounting /sys/firmware/efi/efivars". I'm not sure if this is related, however it's possible that my problem is not a zfsonlinux issue.

**Comment by earsdown** [comment]
I can reproduce this bug using Ubuntu 14.04 on AWS EC2 with EBS volumes.

Steps:
1. Spin up an Ubuntu 14.04 instance in EC2 and make sure it's fully patched
2. Attach 4 EBS SSD volumes (gp2) to sdf, sdg, sdh & sdi
3. Run these commands to set up ZFS the way I have
```
apt-add-repository -y ppa:zfs-native/stable
apt-get update && apt-get -y install ubuntu-zfs
modprobe zfs
for d in /dev/xvdf /dev/xvdg /dev/xvdh /dev/xvdi; do
  parted -s $d unit s mklabel gpt mkpart primary zfs 2048s 100%
done
zpool create bug2944 /dev/disk/by-partuuid/*
zpool status bug2944
```
Confirm that the zpool was created with devices from /dev/disk/by-partuuid/* and not from /dev/xvd*.

Finally, reboot and check zpool status again. When I do this, the zpool is made up of /dev/xvd* devices, rather than from the /dev/disk/by-partuuid/* devices. Of course, I can always export and import, but would be quite awkward. 

Any suggestions?

**Comment by lukeiamyourfather** [comment]
Same issue here as described above where devices imported or created with /dev/disk/by-id go to /dev/sdX after a reboot. This is on Ubuntu 14.04 LTS. If it makes any difference the hardware is SSG-6048R-E1CR24L which has LSI 3008 (IT mode). Interestingly the logs and cache devices show up fine with /dev/disk/by-id and are on a different controller on the motherboard.

**Comment by q-ray** [comment]
Same here. I installed a new Ubuntu 14.04 LTS server because my boot drive crashed. Installed zfs and imported the existing pools by id. Zpool status showed the pools correct with disks by id. After reboot zpool stsus showed the disk with sdx.
I tried zpool export/import several times. Same result on every try :(

**Comment by hybridchemistry** [comment]
Just wanted to chime in that this has been an issue with my Ubuntu 14.04.3 LTS and Perc H310/LSI SAS 9211-8i that have been flashed to IT mode. If a fix is coming for Ubuntu 15, I suppose I can hang on... has anyone else had success with the device wait script?

**Comment by joelross** [comment]
Also met this problem (IBM M1015 reflashed to P16 IT 9211-8i) on a Debian Wheezy derivative (OpenMediaVault 2.1.15). Using option `USE_DISK_BY_ID` in file `/etc/default/zfs` as advised in [this comment](https://github.com/zfsonlinux/zfs/issues/3756#issuecomment-141706660) solved it for me, but I'm not sure that it doesn't juste "hide" the real issue.

By the way, I tried to use P19 on my IBM M1015/LSI 9211-8i using Ubuntu 14.04 (Linux Mint 17 actually). It resulted in tons of UDMA CRC errors, so I went back to P16. It was about 1 year ago and I didn't touch the drivers (to be honest, I didn't know there were newer drivers available).

**Comment by weitzj** [comment]
Same for me. stock Ubuntu 16.04 on a Dell T110 ii. Drives attached to the Dell Perc S110 which might actually be an LSI card

**Comment by behlendorf** [comment]
This was recently fixed in master with commit 325414e483a7858c1d10fb30cefe5749207098f4, the plan is to backport it for the next point release.

**Comment by brando56894** [comment]
I just switched from Ubuntu 14.04 LTS to Arch and ran into this bug also, I too have an LSI HBA flashed to IT mode. Deleting /etc/zfs/zpool.cache and rebooting allows everything to be mounted again.

**Comment by Ralithune** [comment]
Behlendorf - Is this fix back ported?  How can I tell?

**Comment by behlendorf** [comment]
@Ralithune yes, currently it's available as part of PR #4605 from @nedbass .  It will be merged in to the release branch with perhaps a few small tweaks.  But if you want to test it sooner it's there.

**Comment by Ralithune** [comment]
@behlendorf I'm totally new to using git to deploy major software like this - is there a how-to on how to install the release candidate?  I'm also using Ubutnu 14.04.X, and the release candidate appears to be failing the checks for that?

**Comment by behlendorf** [comment]
@Ralithune there's not as much documentation as I'd like but we're working on it.

For testing purposes you should be able to follow the [guide](https://github.com/zfsonlinux/zfs/wiki/Building-ZFS) on the wiki for building ZFS with a few small modifications.

* Add @nedbass as a Git remote and checkout his branch from the PR for the build.
* After building you can use the  Use the `zfs.sh` script to unload the existing kernel modules and load the ones you just built.
* Run the `zpool export -a` then `zpool import <pool>` from cmd/zpool directory to verify the vdev names are preserved as expected.

Alternately, if you can wait a few days we should have it tagged.

**Comment by behlendorf** [comment]
Closing.  Addressed in the 0.6.5.7 release which was just tagged.
